Extension/lib/filter/rules/css-filter-rule.js

```diff
@@ -31,30 +31,126 @@ var FilterRule = require('../../../lib/filter/rules/base-filter-rule').FilterRul
-	if (indexOfMask > 0) {
-		// domains are specified, parsing
-		var domains = rule.substring(0, indexOfMask);
-		this.loadDomains(domains);
-	}
-    
-    this.cssSelector = rule.substring(indexOfMask + mask.length);
+    FilterRule.call(this, rule, filterId);
+
+    var isInjectRule = StringUtils.contains(rule, FilterRule.MASK_CSS_INJECT_RULE) || StringUtils.contains(rule, FilterRule.MASK_CSS_EXCEPTION_INJECT_RULE);
+    if (isInjectRule) {
+        this.isInjectRule = isInjectRule;
+    }
+
+    var mask;
+    if (isInjectRule) {
+        this.whiteListRule = StringUtils.contains(rule, FilterRule.MASK_CSS_EXCEPTION_INJECT_RULE);
+        mask = this.whiteListRule ? FilterRule.MASK_CSS_EXCEPTION_INJECT_RULE : FilterRule.MASK_CSS_INJECT_RULE;
+    } else {
+        this.whiteListRule = StringUtils.contains(rule, FilterRule.MASK_CSS_EXCEPTION_RULE);
+        mask = this.whiteListRule ? FilterRule.MASK_CSS_EXCEPTION_RULE : FilterRule.MASK_CSS_RULE;
+    }
+
+    var indexOfMask = rule.indexOf(mask);
+    if (indexOfMask > 0) {
+        // domains are specified, parsing
+        var domains = rule.substring(0, indexOfMask);
+        this.loadDomains(domains);
+    }
+
+    var cssContent = rule.substring(indexOfMask + mask.length);
+
+    if (!isInjectRule) {
+        var pseudoClass = parsePseudoClass(cssContent);
+        if (pseudoClass != null && ":style" == pseudoClass.name) {
+            isInjectRule = true;
+            cssContent = convertCssInjectionRule(pseudoClass, cssContent);
+        } else if (pseudoClass != null && CssFilterRule.SUPPORTED_PSEUDO_CLASSES.indexOf(pseudoClass.name) < 0) {
+            throw new Error("Unknown pseudo class: " + cssContent);
+        }
+    }
+
+    this.isInjectRule = isInjectRule;
+    this.cssSelector = cssContent;
 };
 
-CssFilterRule.prototype = Object.create(FilterRule.prototype);
\ No newline at end of file
+CssFilterRule.prototype = Object.create(FilterRule.prototype);
+
+/**
+ * The problem with pseudo-classes is that any unknown pseudo-class makes browser ignore the whole CSS rule,
+ * which contains a lot more selectors. So, if CSS selector contains a pseudo-class, we should try to validate it.
+ * <p>
+ * One more problem with pseudo-classes is that they are actively used in uBlock, hence it may mess AG styles.
+ */
+CssFilterRule.SUPPORTED_PSEUDO_CLASSES = [":active",
+    ":checked", ":disabled", ":empty", ":enabled", ":first-child", ":first-of-type",
+    ":focus", ":hover", ":in-range", ":invalid", ":lang", ":last-child", ":last-of-type",
+    ":link", ":not", ":nth-child", ":nth-last-child", ":nth-last-of-type", ":nth-of-type",
+    ":only-child", ":only-of-type", ":optional", ":out-of-range", ":read-only",
+    ":read-write", ":required", ":root", ":target", ":valid", ":visited"];
+
+
+/**
+ * Parses first pseudo class from the specified CSS selector
+ *
+ * @param selector
+ * @returns {*} first PseudoClass found or null
+ */
+function parsePseudoClass(selector) {

```

Comment:
I think having it in the global scope is not very good. However we cannot simply move it inside constructor.

Maybe we should do something like this:
```
exports.CssFilterRule = (function() {
    var privateFunction = function() {};
    var constructor = function(rule, filterId) {
        // blahblah
    };
    return constructor;  
})();
```